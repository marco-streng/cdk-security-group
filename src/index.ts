import * as ec2 from "@aws-cdk/aws-ec2";
import { Construct } from "@aws-cdk/core";
import * as cr from "@aws-cdk/custom-resources";

export interface SecurityGroupProps extends ec2.SecurityGroupProps {}

export class SecurityGroup extends ec2.SecurityGroup {
  /**
   * An attribute that represents the security group name.
   */
  public readonly securityGroupName: string;
  constructor(scope: Construct, id: string, props: SecurityGroupProps) {
    super(scope, id, props);

    const securityGroup = new cr.AwsCustomResource(this, "DescribeSGCustomResource", {
      onCreate: {
        service: "EC2",
        action: "describeSecurityGroups", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/EC2.html#describeSecurityGroups-property
        parameters: {
          GroupIds: [this.securityGroupId],
        },
        physicalResourceId: cr.PhysicalResourceId.of(this.securityGroupId),
      },
      policy: cr.AwsCustomResourcePolicy.fromSdkCalls({
        resources: cr.AwsCustomResourcePolicy.ANY_RESOURCE,
      }),
    });

    this.securityGroupName = securityGroup.getResponseField("SecurityGroups.0.GroupName");
  }
}
