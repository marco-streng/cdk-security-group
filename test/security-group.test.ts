import { Template } from "@aws-cdk/assertions";
import { Vpc } from "@aws-cdk/aws-ec2";
import { Stack } from "@aws-cdk/core";
import { SecurityGroup } from "../src";

describe("SecurityGroup", () => {
  it("should have named security group", () => {
    // Given
    const stack = new Stack();
    const vpc = new Vpc(stack, "vpc");
    // When
    new SecurityGroup(stack, "id", {
      vpc: vpc,
      securityGroupName: "pepperize",
    });
    // Then
    const template = Template.fromStack(stack);
    template.hasResource("AWS::EC2::SecurityGroup", {});
    template.hasResourceProperties("AWS::EC2::SecurityGroup", {
      GroupName: "pepperize",
    });
  });

  it("should match snapshot", () => {
    // Given
    const stack = new Stack();
    const vpc = new Vpc(stack, "vpc");
    // When
    new SecurityGroup(stack, "id", { vpc: vpc, securityGroupName: "pepperize" });
    // Then
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
