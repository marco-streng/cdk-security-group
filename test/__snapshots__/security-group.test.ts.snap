// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecurityGroup should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8dArtifactHash6BDDF757": Object {
      "Description": "Artifact hash for asset \\"70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8d\\"",
      "Type": "String",
    },
    "AssetParameters70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8dS3BucketF0537962": Object {
      "Description": "S3 bucket for asset \\"70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8d\\"",
      "Type": "String",
    },
    "AssetParameters70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8dS3VersionKey46386A78": Object {
      "Description": "S3 key for asset version \\"70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8d\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8dS3BucketF0537962",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8dS3VersionKey46386A78",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8dS3VersionKey46386A78",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "id4D1900E5": Object {
      "Properties": Object {
        "GroupDescription": "Default/id",
        "GroupName": "pepperize",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "idDescribeSGCustomResource10A5E3D6": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "idDescribeSGCustomResourceCustomResourcePolicy7FAA6A6B",
      ],
      "Properties": Object {
        "Create": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"EC2\\",\\"action\\":\\"describeSecurityGroups\\",\\"parameters\\":{\\"GroupIds\\":[\\"",
              Object {
                "Fn::GetAtt": Array [
                  "id4D1900E5",
                  "GroupId",
                ],
              },
              "\\"]},\\"physicalResourceId\\":{\\"id\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "id4D1900E5",
                  "GroupId",
                ],
              },
              "\\"}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "idDescribeSGCustomResourceCustomResourcePolicy7FAA6A6B": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ec2:DescribeSecurityGroups",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "idDescribeSGCustomResourceCustomResourcePolicy7FAA6A6B",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "vpcA2121C38": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/vpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "vpcIGWE57CBDCA": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/vpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "vpcPrivateSubnet1DefaultRoute1AA8E2E5": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "vpcPublicSubnet1NATGateway9C16659E",
        },
        "RouteTableId": Object {
          "Ref": "vpcPrivateSubnet1RouteTableB41A48CC",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vpcPrivateSubnet1RouteTableAssociation67945127": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vpcPrivateSubnet1RouteTableB41A48CC",
        },
        "SubnetId": Object {
          "Ref": "vpcPrivateSubnet1Subnet934893E8",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcPrivateSubnet1RouteTableB41A48CC": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/vpc/PrivateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcPrivateSubnet1Subnet934893E8": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.128.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "Name",
            "Value": "Default/vpc/PrivateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcPrivateSubnet2DefaultRouteB0E07F99": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "vpcPublicSubnet2NATGateway9B8AE11A",
        },
        "RouteTableId": Object {
          "Ref": "vpcPrivateSubnet2RouteTable7280F23E",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vpcPrivateSubnet2RouteTable7280F23E": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/vpc/PrivateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcPrivateSubnet2RouteTableAssociation007E94D3": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vpcPrivateSubnet2RouteTable7280F23E",
        },
        "SubnetId": Object {
          "Ref": "vpcPrivateSubnet2Subnet7031C2BA",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcPrivateSubnet2Subnet7031C2BA": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.192.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "Name",
            "Value": "Default/vpc/PrivateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcPublicSubnet1DefaultRoute10708846": Object {
      "DependsOn": Array [
        "vpcVPCGW7984C166",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "vpcIGWE57CBDCA",
        },
        "RouteTableId": Object {
          "Ref": "vpcPublicSubnet1RouteTable48A2DF9B",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vpcPublicSubnet1EIPDA49DCBE": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/vpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "vpcPublicSubnet1NATGateway9C16659E": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "vpcPublicSubnet1EIPDA49DCBE",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "vpcPublicSubnet1Subnet2E65531E",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/vpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "vpcPublicSubnet1RouteTable48A2DF9B": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/vpc/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcPublicSubnet1RouteTableAssociation5D3F4579": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vpcPublicSubnet1RouteTable48A2DF9B",
        },
        "SubnetId": Object {
          "Ref": "vpcPublicSubnet1Subnet2E65531E",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcPublicSubnet1Subnet2E65531E": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "Default/vpc/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcPublicSubnet2DefaultRouteA1EC0F60": Object {
      "DependsOn": Array [
        "vpcVPCGW7984C166",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "vpcIGWE57CBDCA",
        },
        "RouteTableId": Object {
          "Ref": "vpcPublicSubnet2RouteTableEB40D4CB",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vpcPublicSubnet2EIP9B3743B1": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/vpc/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "vpcPublicSubnet2NATGateway9B8AE11A": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "vpcPublicSubnet2EIP9B3743B1",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "vpcPublicSubnet2Subnet009B674F",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/vpc/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "vpcPublicSubnet2RouteTableAssociation21F81B59": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vpcPublicSubnet2RouteTableEB40D4CB",
        },
        "SubnetId": Object {
          "Ref": "vpcPublicSubnet2Subnet009B674F",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcPublicSubnet2RouteTableEB40D4CB": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Default/vpc/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcPublicSubnet2Subnet009B674F": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.64.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "Default/vpc/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcVPCGW7984C166": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "vpcIGWE57CBDCA",
        },
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
}
`;
